core/src/main/java/com/yandex/yoctodb/v1/mutable/segment/V1StoredIndex.java

```diff
@@ -23,29 +24,35 @@
         implements IndexSegment {
     @NotNull
     private final byte[] fieldName;
-    private SortedMap<Integer, UnsignedByteArray> values = new TreeMap<>();
+    private Map<UnsignedByteArray, LinkedList<Integer>> valueDocId = new LinkedHashMap<>();
     private int databaseDocumentsCount = -1;
+    private boolean uniqueValues = true;
+    private int segmentTypeCode;
 
     public V1StoredIndex(
-            @NotNull
-            final String fieldName) {
+            @NotNull final String fieldName) {
         this.fieldName = fieldName.getBytes();
     }
 
     @NotNull
     @Override
     public IndexSegment addDocument(
             final int documentId,
-            @NotNull
-            final Collection<UnsignedByteArray> values) {
+            @NotNull final Collection<UnsignedByteArray> values) {
         if (documentId < 0)
             throw new IllegalArgumentException("Negative document ID");
         if (values.size() != 1)
             throw new IllegalArgumentException("A single value expected");
 
         checkNotFrozen();
 
-        this.values.put(documentId, values.iterator().next());
+        final UnsignedByteArray value = values.iterator().next();
+
+        this.valueDocId.putIfAbsent(value, new LinkedList<>());
```

Comment:
It will be better not to create new `LinkedList` on each `addDocument()` call.
There are fancy `Map` methods like `computeIfAbsent` etc.